python/cudf/cudf/core/udf/typing.py

```diff
@@ -0,0 +1,186 @@
+        return f"MaskedType({self.value_type})"
+
+    def __hash__(self):
+        '''
+        Needed so that numba caches type instances with different
+        `value_type` separately.  
+        '''
+        return self.__repr__().__hash__()
+
+    def unify(self, context, other):
+        '''
+        Logic for sorting out what to do when the UDF conditionally
+        returns a `MaskedType`, an `NAType`, or a literal based off 
+        the data at runtime.
+
+        In this framework, every input column is treated as having
+        type `MaskedType`. Operations like `x + y` are understood 
+        as translating to:
+
+        `Masked(value=x, valid=True) + Masked(value=y, valid=True)`
+
+        This means if the user writes a function such as 
+        def f(x, y):
+            return x + y
+            
+        numba sees this function as:
+        f(x: MaskedType, y: MaskedType) -> MaskedType
+        
+        However if the user writes something like:
+        def f(x, y):
+            if x > 5:
+                return 42
+            else:
+                return x + y
+        
+        numba now sees this as
+        f(x: MaskedType, y: MaskedType) -> MaskedType OR literal 
+        '''
+        
+        # If we have Masked and NA, the output should be a 
+        # MaskedType with the original type as its value_type
+        if isinstance(other, NAType):
+            return MaskedType(self.value_type)
+
+        # if we have MaskedType and Literal, the output should be
+        # determined from the MaskedType.value_type (which is a 
+        # primitive type) and other
+        unified = context.unify_pairs(self.value_type, other)
+        if unified is None:
+            return None
+
+        return MaskedType(unified)
+
+# Tell numba how `MaskedType` is constructed on the backend in terms
+# of primitive things that exist at the LLVM level
+@register_model(MaskedType)
+class MaskedModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        # This struct has two members, a value and a validity
+        # let the type of the `value` field be the same as the 
+        # `value_type` and let `valid` be a boolean 
+        members = [("value", fe_type.value_type), ("valid", types.bool_)]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+class NAType(types.Type):
+    '''
+    A type for handling ops against nulls
+    Exists so we can:
+    1. Teach numba that all occurances of `cudf.NA` are 
+       to be read as instances of this type instead
+    2. Define ops like `if x is cudf.NA` where `x` is of 
+       type `Masked` to mean `if x.valid is False`
+    '''
+    def __init__(self):
+        super().__init__(name="NA")
+
+    def unify(self, context, other):
+        '''
+        Masked  <-> NA works from above
+        Literal <-> NA -> Masked
+        '''
+        if isinstance(other, types.abstract.Literal):
+            return MaskedType(other.literal_type)
```

Comment:
I think here we don't just want to restrict this to literal types - I think any non-masked type should get masked. So perhaps two or three steps would be needed here:

1. If the other is a `MaskedType`, then return `None` here, so this attempt at unification fails, and the `unify` from the `MaskedType` will be subsequently tried.
2. If the other type is also `NA`, then return `self` to make sure that we don't end up with a masked NA (I'm not 100% sure this can happen, and will need to check whether a type can end up in a set with another type the same).
3. Otherwise, return the a masked type that masks the `other` type.